src/pool/srv_cli.c

```diff
@@ -134,3 +139,341 @@ dsc_pool_tgt_reint(const uuid_t uuid, const char *grp,
+			DP_PRE(uuid, cbs), DP_RC(rc));
+		return rc;
+	}
+
+	/*
+	 * Since it's common that the first RPC gets a "not leader" reply with
+	 * a valid hint on who the leader is, let the first backoff be zero.
+	 */
+	rc = d_backoff_seq_init(&backoff_seq, 1 /* nzeros */, 16 /* factor */, 8 /* next (ms) */,
+				1 << 12 /* max (ms) */);
+	D_ASSERTF(rc == 0, DF_PRE": initialize backoff sequence: "DF_RC"\n", DP_PRE(uuid, cbs),
+		  DP_RC(rc));
+
+	/* Retry until the deadline. */
+	for (;;) {
+		crt_endpoint_t		ep;
+		crt_rpc_t	       *rpc;
+		uint32_t		rpc_timeout;
+		uint64_t		t;
+		struct pool_op_out     *out;
+		uint32_t		backoff = d_backoff_seq_next(&backoff_seq);
+
+		ep.ep_grp = NULL;
+		rc = rsvc_client_choose(&client, &ep);
+		if (rc != 0) {
+			D_ERROR(DF_PRE": choose pool service replica: "DF_RC"\n", DP_PRE(uuid, cbs),
+				DP_RC(rc));
+			break;
+		}
+
+		rc = pool_req_create(info->dmi_ctx, &ep, cbs->pscc_op, &rpc);
+		if (rc != 0) {
+			D_ERROR(DF_PRE": create RPC: "DF_RC"\n", DP_PRE(uuid, cbs), DP_RC(rc));
+			break;
+		}
+
+		rc = cbs->pscc_init(uuid, rpc, arg);
+		if (rc != 0) {
+			D_ERROR(DF_PRE": initialize RPC: "DF_RC"\n", DP_PRE(uuid, cbs), DP_RC(rc));
+			crt_req_decref(rpc);
+			break;
+		}
+
+		/*
+		 * Cap the RPC timeout according to the deadline. It's unlikely
+		 * that we have passed the deadline here; comparing t with
+		 * deadline is just to make the subtraction below obviously
+		 * underflow-free.
+		 */
+		t = daos_getmtime_coarse();
+		if (t > deadline) {
+			cbs->pscc_fini(uuid, rpc, arg);
+			crt_req_decref(rpc);
+			goto time_out;
+		}
+		rc = crt_req_get_timeout(rpc, &rpc_timeout);
+		D_ASSERTF(rc == 0, DF_PRE": get RPC timeout: "DF_RC"\n", DP_PRE(uuid, cbs),
+			  DP_RC(rc));
+		if (t + rpc_timeout * 1000 > deadline) {
+			/* We know that deadline >= t here. See above. */
+			rpc_timeout = (deadline - t) / 1000;
+			/*
+			 * If the RPC timeout becomes less than 1 s, just time
+			 * out the call.
+			 */
+			if (rpc_timeout < 1) {
+				cbs->pscc_fini(uuid, rpc, arg);
+				crt_req_decref(rpc);
+				goto time_out;
+			}
+			rc = crt_req_set_timeout(rpc, rpc_timeout);
+			D_ASSERTF(rc == 0, DF_PRE": set RPC timeout: "DF_RC"\n", DP_PRE(uuid, cbs),
+				  DP_RC(rc));
+		}
+
+		rc = dss_rpc_send(rpc);
+
+		out = crt_reply_get(rpc);
+		rc = rsvc_client_complete_rpc(&client, &ep, rc, rc == 0 ? out->po_rc : 0,
+					      rc == 0 ? &out->po_hint : NULL);
+		if (rc == RSVC_CLIENT_PROCEED && !daos_rpc_retryable_rc(out->po_rc)) {
+			rc = cbs->pscc_consume(uuid, rpc, arg);
+			D_DEBUG(DB_TRACE, DF_PRE": consume: %d\n", DP_PRE(uuid, cbs), rc);
+			if (rc == DSC_POOL_SVC_CALL_AGAIN_NOW) {
+				backoff = 0;
+			} else if (rc != DSC_POOL_SVC_CALL_AGAIN) {
+				cbs->pscc_fini(uuid, rpc, arg);
+				crt_req_decref(rpc);
+				break;
+			}
+		}
+
+		cbs->pscc_fini(uuid, rpc, arg);
+		crt_req_decref(rpc);
+
+		t = daos_getmtime_coarse();
+		if (t + backoff >= deadline) {
```

Comment:
I think if we reach that close to the deadline, carrying out the steps at the top of the loop may not matter too much. When I tried applying the suggestion, I found that returning before reaching the deadline were actually undesirable with the current control plane code. Please see the new comment in the revised PR. So in the end the suggestion has been very helpful. :)

Reply:
Nice! I see you have added a conditional explicit sleep in the logic at the time_out label to prevent control plane logic from retrying so close to the deadline. Looks good.